[WIP] PEP8

mottosso opened the pull request:
Hi Miquel,

As discussed in #94, in this PR I'll begin refactoring code for compliance with PEP8.

The goal is to..

- Enhance maintainability
- Encourage open source contributions

Maintainability is enhanced by improved readability and the ability to use interactive linters when writing new code. Linters will enable new code to remain consistent with old code such that new developers are (1) familiar with old code and (2) won't need to ask about how to format new code.

At this point, Travis will find that tests still break. These can be added in parallel to this PR if necessary. Ideally, we'd have Travis run through the code during refactoring to check for syntax problems.

Code Climate (CC) will run in parallel to Travis and tell us how we're doing. CC will do what interactive linters do, and run the code through flake8 and others. Once CC is happy, we know we've done a good job! But, even there, there are a few rules worth breaking that we'll chat about and provide a proper motivation for our future self and future others reading the code anew.

## Rules

I'll edit this paragraph over time to reflect rules applied to the overall PR.

| Rule                           | Description
|:-------------------------------|:--------------
| Use mixedCase                  | camel_case is more Pythonic, but mgear builds upon an API (Maya) that uses mixedCase so it's either mixing two conventions, or biting the bullet.
| Napoleon Docstring             | Once/if we run mgear through Sphinx, [Napoleon](http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) is a good compromise between Sphinx functionality and code readability.
| No newline in docstring title  |  See [here](https://github.com/miquelcampos/mgear/pull/100/files#diff-383c9437e651875812fd1881c02a356cR9)
| No renaming of imports         | Right: `from mgear.maya import shifter`                                         Wrong: `import mgear.maya.shifter as sg`
| No license in every file       | Clutters every file. This is only relevant when a file is distributed independently, but mgear is always distributed as a whole, including the LICENSE file.

mottosso commented:
(Ps. the additional commits are from #93 which we should merge first)

miquelcampos commented:
Hi Marcus,

This looks good for me.
I have been using Napoleon already (or at less trying)

I have to push some scons changes and later I will  merge CC PR

mottosso commented:
I've updated the PR to the latest `develop` by rebasing; which means that if you cloned this PR already, you'll need to reset it, or re-checkout. I rebased this time to keep the start of the PR clean, will merge in the future so that no checkout breaks.

Also, you can push to this PR @miquelcampos, you do it by adding mottosso/pep8 to your local git repo, and pushing directly to it. You have write permissions for it, as it's part of this PR in your account name. That way we can work on the same PR together!

miquelcampos commented:
@mottosso Oh cool!  I didn't know that you can do that.
I will try now, lets see what I break

miquelcampos commented:
Sweet!

miquelcampos commented:
Question.. how to split the this task?   
Do you want me to do a specific folders or files?

mottosso commented:
> Do you want me to do a specific folders or files?

Good question. How about we spit it 50/50 initially, until one of us finished the first half? Since you know the code, why don't you pick a half to dedicate to yourself, and then I'll take the other half? :)

miquelcampos commented:
OK,
Let me start with Shifter .   
If you want to start with the scripts base folder, maya folder and mgear folder. 

Later we can do the folder (rigbits, synoptic, simpleRig)

mottosso commented:
👍 

For interactive linting, I'm using Sublime Text with Anaconda, which is using PyFlakes and PyLint. I think it should align with Code Climate, but if in doubt, let's use Code Climate as a common ground and configure it as necessary.

miquelcampos commented:
Ah ok! I have sublime linter wiht Flake8 and pep8 but I suspect my configuration is not correct.
I will change to anaconda

mottosso commented:
Now that I think of it, it might be a better idea to use PyFlakes and PyLint as a common ground, as it would mean users could contribute code whilst using default settings on their linters. The alternative is them either having to configure their linters specifically for this project, or find out once the submit a PR that rules are broken.

We could configure CC to align with Anaconda, for example. Let's see how it goes.

miquelcampos commented:
right now just researching anaconda configuration. Do you need to configure anything to get PyFlakes and PyLint in anaconda?

miquelcampos commented:
Just checking your first commit for the PR
 multiple import in 1 line? https://www.python.org/dev/peps/pep-0008/#imports

mottosso commented:
> Do you need to configure anything to get PyFlakes and PyLint in anaconda?

Mine is unconfigured, defaults should be ok.

> Just checking your first commit for the PR, multiple import in 1 line?

Are you getting a linting error on that? See the next line in your link.

> It's okay to say this though: 
> from subprocess import Popen, PIPE

Basically, anything starting with `from` is OK (which we should be using a lot more, as it means (1) you can bundle mgear in another Python package and most importantly (2) it means there is no risk of mgear picking up modules from a different mgear distribution on a users computer)

miquelcampos commented:
No, no , I am not getting any error from that. Just asking for confirmation :)

miquelcampos commented:
Anaconda: Rename name under cursor tool is crashing Sublime text 3 .  Are you using this tool?

mottosso commented:
I'm not, no.

mottosso commented on scripts/mgear/maya/shifter/__init__.py:
```diff
@@ -1,76 +1,38 @@
 import mgear.maya.utils
-from mgear.maya.shifter.guide import RigGuide
-from mgear.maya.shifter.guide import helperSlots
+from mgear.maya.shifter.guide import RigGuide, helperSlots
```
These can be reduced to..

```python
from .guide import RigGuide
```

Though I'd suggest we keep the namespace, and make it..

```python
from . import guide
```

miquelcampos replied:
something like this?
'''
from . import guide
from guide import RigGuide, helperSlots
'''

mottosso replied:
I was thinking more along these lines..

```python
from . import guide
...
self.guide = guide.RigGuide()
```

If you stick with namespaces, you can then also reduce variable names. For example..

```python
self.guide = guide.Rig()
```

See how it reduces the duplicate mention of "guide"?

miquelcampos replied:
Not sure if I understand this correctly.  Do you mind to change this file for a full example? Just to be sure.  Thanks!

mottosso replied:
Sure, no problem. I'll take a look at this tomorrow morning.

mottosso replied:
Ok, I've pushed an example.

1. The namespace of the `component` and `guide` modules are maintained
2. Names are shortened to take into consideration the namespace
3. Backwards compatibility is maintained via aliases

This highlights an issue with the current names of the modules, `guide` is a module but also an often used variable name, as is `component`. I can't say there's a straightforward solution, except to avoid nouns for module names, and reserve those for variables. For example `utility.py` and `io.py` is unlikely to conflict with variables, whereas `object` and `item` likely are.

The shortening isn't as important as the overall PEP8 changes, so we could delay those until we're done with that.

miquelcampos replied:
Thank you,  I am checking now the changes

miquelcampos replied:
Oh I get it now! 
So we use simple names but always inside the namespace. So is easily identifiable from where is coming.
While creating the alias just in case for back compatibility. 
looks good for me. thanks!

mottosso replied:
Cool, glad you like it. :)

Once we get to defining an API for mgear, this will become even crisper and easily readable. I'll pop open an issue with some thoughts about that later today.

mottosso commented on scripts/mgear/maya/shifter/__init__.py:
```diff
@@ -341,53 +317,51 @@ def finalize(self):
     def addCtl(self, parent, name, m, color, icon, **kwargs):
-        """
-        Create the control and apply the shape, if this is alrealdy stored
+        """Create the control and apply the shape, if this is alrealdy stored
```
First line of a docstring should be a title, or summary. It shouldn't continue on the next line. In Sphinx, this will be used as a header or summary for the subsequent lines, which will become the body.

- [Examples](http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

miquelcampos replied:
Thanks for the notes.   I will continue tomorrow. :)

mottosso commented on scripts/mgear/maya/shifter/gameTools.py:
```diff
@@ -187,61 +183,65 @@ def exportAssetAssembly(name, rigTopNode, meshTopNode, path, postScript=None):
-
-
-    #pot script
+    # pot script
```
Pot or Post? :)

miquelcampos replied:
should be post, this a typo. 
However this tool is not finished. I have put this on hold until finish this refactor

miquelcampos commented:
@mottosso Silly question about splitting lines:
![image](https://user-images.githubusercontent.com/1050212/32883277-400dcdb8-cafa-11e7-844e-8272762adf3c.png)

is correct to do this?
![image](https://user-images.githubusercontent.com/1050212/32883301-56cbee7c-cafa-11e7-80de-792a097e0cfe.png)

mottosso commented:
We'll have to work on finding more succinct variable names to avoid this problem more generally, but in this particular case I would probably assign a variable to the checkbox.

```python
tab = self.guideSettingsTab
checkbox = tab.classicChannelNames_checkBox
checkbox.connect(...)
```

miquelcampos commented:
Yep this looks cleaner. Thanks for the tip!

mottosso commented:
Something I typically do to separate what is a variable and what is a widget is to put widgets in a top-level dictionary.

```python
self.widgets = {
  "classicChannelNames_checkBox": classicChannelNames_checkBox,
  ...
}
```

Then you could organise things by type, if necessary.

```python
self.widgets = {
  "checkbox": {
    "classicChannelNames": classicChannelNames_checkBox,
    ...
  }
}
```

To be used like this..

```python
self.widgets["checkbox"]["classicChannelNames"].stateChanged.connect(...)
```

That way you can shorten variable names, and keep widgets in one place. Then you can do the same for state, like `self.state["isRunning"]` etc.

miquelcampos commented:
Really good idea! I am learning a lot of new tricks today :dancing_men:

miquelcampos commented:
2de9a4a Just pushed to continue from home

miquelcampos commented:
Hi @mottosso I am thinking that we should not refactor to PEP8 any module generated by compiling a .ui file from QT designer. 
What do you think?

btw: after a few files refactored I realized very well about the inconvenience of using inappropriate names for the modules and variables. I should remember this for 3.X version

mottosso commented:
Yes, good point. As we go, I'll propose we shift from .ui files to hand-crafted widgets, in which case this won't be a problem anymore.

You can add the generated files to the `.codeclimate.yml` exclusion list, that way it won't bother keeping track of those.

miquelcampos commented:
Ok, I will check now the yml file.

Why do you want to change to hand-crafted files? I guess depending of the tool maybe is an advantage, but for simple interface I am no sure. 
Anyway, you are the QT expert  here :)

miquelcampos commented:
Hello @mottosso 

After a few files refactored. I have been thinking that the code has become a little harder to read because the max line width of 79 characters and my lack of clear style of how to arrange the lines that doesn't fit in the width.

So I decided to follow this  rules:

If we have a line like this:
`applyop.aimCns(self.ref_base,  self.squash_ctl,axis="yx", wupType=2,  wupVector=[1, 0, 0], wupObject=self.ctl, maintainOffset=False)`

1st preference Try to indent all arguments to the parenthesis:
```
applyop.aimCns(self.ref_base,
                       self.squash_ctl,
                       axis="yx",
                       wupType=2,
                       wupVector=[1, 0, 0],
                       wupObject=self.ctl,
                       maintainOffset=False)
```

2nd preference: indent all the the next indentation of the line
```
applyop.aimCns(
    self.ref_base,
    self.squash_ctl,
    axis="yx",
     wupType=2,
     wupVector=[1, 0, 0],
     wupObject=self.ctl,
     maintainOffset=False)
```

avoid more than one argument if we have to indent
Example of Dont's
```
applyop.aimCns(self.ref_base, self.squash_ctl, axis="yx",
                       wupType=2, wupVector=[1, 0, 0], wupObject=self.ctl,
                       maintainOffset=False)
```

What do you think?

I will prepare some documentation about coding style and collaboration workflow after finish this refactor.

mottosso commented:
I think that sounds great. :)

> I will prepare some documentation about coding style and collaboration workflow after finish this refactor.

It can be tricky to remember all the minutia once you're done, I'd suggest you keep a little bullet-list of rules, like the one you just mentioned. Sometimes, a bullet list is all it takes for "documentation".

- Long argument line break, align with call tail
```python
# No
function(arg1, arg2,
         kwarg=False, kwarg2=True)

# Yes
function(arg1,
         arg2,
         kwarg=False,
         kwarg2=True)

# OK
function(
    arg1,
    arg2,
    kwarg=False,
    kwarg2=True)

```

miquelcampos commented:
Hi @mottosso I have finish with Shifter. (I have tried to test by building every component, but I  don't discard some bugs will show up later)
I  will continue now with the core library under mgear and maya folder

sporty commented:
+1

mottosso commented:
That's excellent @miquelcampos, I expected to put in more time on this but have overloaded myself with tasks for the past few days. I'll have a look at the `mGear_*` modules in the `scripts/` directory today.

miquelcampos commented:
Thanks @mottosso!  This ones are provably the most dirty in terms of code. I hope don't scare you. :P

mottosso commented:
Yes, once we've pepified things, I'll propose we discard these in favour of an `install` function, and embed most of it into the mgear package.

miquelcampos commented:
Yep, you are right. Indeed some of them are already move to a better organisation. Like the "mGear_riggingTools.py"

what is an `install `function?

mottosso commented:
Btw, about [this one](), I removed a lot of unnecessary use of `partial` which shouldn't affect how they work, but we should keep an eye out for any broken buttons. Then, once we're done, I'll propose a factory function to simplify that module a lot.

```python
# E.g.
mgear.create_menu((
  ("menuItem1", command1),
  ("menuItem2", command2),
  ...
))
```

miquelcampos commented:
Oh! I see.
ones of the ideas I have is to add the input to the menus in the  modules, so when we split it, we will get a dynamic menu population based on which  module are in the package.
No more hardcoded menu

miquelcampos commented:
I think some of the partial functions still needed, because default arguments are not the same as it was in the partial call. 
Provably this is bad design from my part. 
Anyway I will review this :)

mottosso commented:
> No more hardcoded menu

Yeah, that's a good idea.

> because default arguments are not the same as it was in the partial call.

I've had a look at a handful of default arguments and the ones I checked were the same. Mostly the ones with `None, None` in them, for example. But, I wouldn't be surprised if a few were needed.

miquelcampos commented:
Thanks for the "Minor tweaks to Shifter" and the CONTRIBUTING.md update
I will try to follow same pattern

mottosso commented:
No problem. The CONTRIBUTING docs are growing quite large, and frankly if it's too long nobody will bother reading it.

One way to tackle that is to have a short summary on-top, with links to descriptions below. Like a "Table of contents".

**Table of Contents**

- [Declare all members in `__init__`]()
- [Use relative imports]()
- [How to line break long arguments]()
- ...

miquelcampos commented:
Is  good idea!  I will review the CONTRIBUTING.md this weekend

miquelcampos commented:
@mottosso Quick question about docstring :

I found that you are using `Arguments:` instead of `Args:` However in http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html is using `Args:`

I guess is a minor thing but I wonder if there is any advantage?  like better readability on the generated sphinx docs?

miquelcampos commented:
@mottosso I will continue now with "rigbits" folder

btw: I still have to update the CONTRIBUTING.md index

miquelcampos commented:
@mottosso I have finished rigbits. (I will do some test tomorrow to ensure everything still working )
Meanwhile I will continue with Simple rig and synoptic

miquelcampos commented:
Hi @mottosso I think the initial PEP8 refactor is finished :)
I did some test building rigs (some with many custom steps) and looks like is not breaking
Am I missing something?  merge?

mottosso commented:
Sweet :) I think let's merge as-is and keep an eye out for non-PEP things as we go. Great job!

miquelcampos commented:
Done! :)
